mono/mini/mini-exceptions.c

```diff
@@ -2749,10 +2773,15 @@ build_stack_trace (struct _Unwind_Context *frame_ctx, void *state)
 static void
 throw_exception (MonoObject *ex, gboolean rethrow)
 {
+	MonoError error;
 	MonoJitTlsData *jit_tls = mono_get_jit_tls ();
 	MonoException *mono_ex;
+	gboolean isinst;
 
-	if (!mono_object_isinst (ex, mono_defaults.exception_class))
+	isinst = mono_object_isinst_checked (ex, mono_defaults.exception_class, &error) == ex;
+	mono_error_raise_exception (&error); /* FIXME don't raise here */
```

Comment:
The behavior should not be to throw, it should be to either assert or convert to an ExecutionEngineException.